storage: add wrapper around rdkafka apis to prevent dropping in tokio tasks

guswynn opened the pull request:
Turns out this was non-trivial, as things like `BaseProducer` are `Clone`, which means we need to duplicate a lot of code to ensure people don't wrap it. There are also some ordering issues about wrappers like `PartitionQueue` that are hard to reason about, but the comments should explain them!


I ensured that the repro (https://github.com/MaterializeInc/materialize/commit/cec7fb80544459c4504ef7c42de65eb35aac64b8) was fixed.

### Motivation

  * This PR fixes a recognized bug.
This fixes the root cause of incident-92, which caused a hard deadlock in our tokio runtime


### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

mjibson reviewed (commented):
Adapter parts lgtm.

benesch commented:
Mentioned this in one of the Slack threads, but I think we should consider just patching our fork of rdkafka! If I understand things correctly, I believe that would be less net complexity!

guswynn commented on src/storage-types/src/connections.rs:
```diff
@@ -452,17 +458,187 @@ impl<C: ConnectionAccess> KafkaConnection<C> {
+    /// `BaseConsumer`.
+    ///
+    // TODO(guswynn): this is probably still unsafe; A `RdkafkaWrapper<PartitionQueue>` that
+    // outlives the original `BaseConsumer` it came from will drop the `BaseConsumer` on the
```
Oh i had fixed this, the comment is outdated

guswynn commented on src/storage-types/src/connections.rs:
```diff
@@ -452,17 +458,187 @@ impl<C: ConnectionAccess> KafkaConnection<C> {
+    // outlives the original `BaseConsumer` it came from will drop the `BaseConsumer` on the
+    // current thread. This scenario is likely rare, might already cause a deadlock in `rdkafka`,
+    // and requires changing the `rdkafka` crate to resolve. We should figure out how to fix this.
+    pub fn split_partition_queue_for_sources<F>(
```
No, because we need to wrap the `PartitionQueue`, and this takes `Arc<Self>`; we don't want to deref to an `Arc<Self>` to make this callable; that would allow the user to clone the underlying consumer

petrosagg replied:
I didn't follow that, what's the problem with this:

```diff
diff --git a/src/storage-types/src/connections.rs b/src/storage-types/src/connections.rs
index 6c6a6dcc6d..4a99d5f90d 100644
--- a/src/storage-types/src/connections.rs
+++ b/src/storage-types/src/connections.rs
@@ -518,30 +518,31 @@ impl<C: ConsumerContext + Send + 'static> RdkafkaWrapper<BaseConsumer<C>> {
     // outlives the original `BaseConsumer` it came from will drop the `BaseConsumer` on the
     // current thread. This scenario is likely rare, might already cause a deadlock in `rdkafka`,
     // and requires changing the `rdkafka` crate to resolve. We should figure out how to fix this.
-    pub fn split_partition_queue_for_sources<F>(
+    pub fn split_partition_queue(
         &self,
         topic: &str,
         partition: i32,
-        non_empty_callback: F,
-    ) -> Option<RdkafkaWrapperPartitionQueue<C>>
-    where
-        F: Fn() + Send + Sync + 'static,
-    {
+    ) -> Option<RdkafkaWrapperPartitionQueue<C>> {
         let consumer = self.clone();
         self.inner
             .as_ref()
             .unwrap()
             .split_partition_queue(topic, partition)
-            .map(|mut q| {
-                q.set_nonempty_callback(non_empty_callback);
-
-                RdkafkaWrapperPartitionQueue {
-                    partition_queue: q,
-                    _consumer: consumer,
-                }
+            .map(|q| RdkafkaWrapperPartitionQueue {
+                partition_queue: q,
+                _consumer: consumer,
             })
     }
 }
+impl<C: ConsumerContext + Send + 'static> RdkafkaWrapperPartitionQueue<C> {
+    pub fn set_nonempty_callback<F>(&mut self, non_empty_callback: F)
+    where
+        F: Fn() + Send + Sync + 'static,
+    {
+        self.partition_queue
+            .set_nonempty_callback(non_empty_callback);
+    }
+}
 
 // We don't implement `Deref` generically over any `T`. This is because some `rdkafka` apis
 // (namely, `BaseProducer`) are `: Clone`. Exposing those by `Deref`-ing them would allow the user
diff --git a/src/storage/src/source/kafka.rs b/src/storage/src/source/kafka.rs
index 07347373e9..4b72625c53 100644
--- a/src/storage/src/source/kafka.rs
+++ b/src/storage/src/source/kafka.rs
@@ -776,19 +776,19 @@ impl KafkaSourceReader {
         for pc in &mut self.partition_consumers {
             pc.partition_queue = self
                 .consumer
-                .split_partition_queue_for_sources(&self.topic_name, pc.pid, {
-                    let context = Arc::clone(&context);
-                    move || context.inner().activate()
-                })
+                .split_partition_queue(&self.topic_name, pc.pid)
                 .expect("partition known to be valid");
+            let context = Arc::clone(&context);
+            pc.partition_queue
+                .set_nonempty_callback(move || context.inner().activate());
         }
 
-        let partition_queue = self
+        let mut partition_queue = self
             .consumer
-            .split_partition_queue_for_sources(&self.topic_name, partition_id, move || {
-                context.inner().activate()
-            })
+            .split_partition_queue(&self.topic_name, partition_id)
             .expect("partition known to be valid");
+        partition_queue.set_nonempty_callback(move || context.inner().activate());
+
         self.partition_consumers.push(PartitionConsumer::new(
             partition_id,
             partition_queue,

```

guswynn commented on src/storage-types/src/connections.rs:
```diff
@@ -452,17 +458,187 @@ impl<C: ConnectionAccess> KafkaConnection<C> {
+
+impl<T: Send + 'static> Drop for RdkafkaWrapper<T> {
+    fn drop(&mut self) {
+        // TODO(guswynn): we would prefer to use `spawn_blocking` here, but we are unsure if all
```
I like this; one thing to note about the difference between `spawn_blocking` and spawning a thread is that if we end up with a deadlock in the future (somehow), the `spawn_blocking` case will cause work to be enqueued on the tokio blocking pool, waiting to be run, which starves the rest of the system. The thread spawning case just leaks.

I think leaking is slightly less risky, but perhaps this concern is purely philosophical

guswynn commented:
I think doing it in rdkafka would be a much bigger change. I don’t feel comfortable just calling the `rd_kafka_destroy` functions on a separate thread/spawn_blocking, as many types in `rust-rdkafka` have multiple fields, and calling a c ffi function after (on a different thread) the other fields have been fully dropped seems like a recipe for disaster. This means everything (at least everything that we use) has to be wrapped up in Option’s, so they can be taken out and moved onto other threads to be dropped.

That would affect a huge amount of code, whereas doing it in the mz repo lets us just wrap things up in options in one place (we really only ever make clients with this one method (except in tests), and we ban at least some misuse with clippy already)

@petrosagg I updated the pr with your suggestions!

petrosagg commented:
@guswynn this conversation seems still unresolved? did you forget to add a commit perhaps? https://github.com/MaterializeInc/materialize/pull/24537/files#r1459387240

benesch commented:
> I think doing it in rdkafka would be a much bigger change. I don’t feel comfortable just calling the `rd_kafka_destroy` functions on a separate thread/spawn_blocking, as many types in `rust-rdkafka` have multiple fields, and calling a c ffi function after (on a different thread) the other fields have been fully dropped seems like a recipe for disaster. This means everything (at least everything that we use) has to be wrapped up in Option’s, so they can be taken out and moved onto other threads to be dropped.
> 
> That would affect a huge amount of code, whereas doing it in the mz repo lets us just wrap things up in options in one place (we really only ever make clients with this one method (except in tests), and we ban at least some misuse with clippy already)

I'm not sure I follow! It seems like a much smaller change to make to rust-rdkafka? Is there a fundamental problem with a diff like the following?

```diff
diff --git a/src/client.rs b/src/client.rs
index d0c6d6fa..c1f9328a 100644
--- a/src/client.rs
+++ b/src/client.rs
@@ -175,7 +175,11 @@ pub struct NativeClient {
 
 unsafe impl KafkaDrop for RDKafka {
     const TYPE: &'static str = "client";
-    const DROP: unsafe extern "C" fn(*mut Self) = rdsys::rd_kafka_destroy;
+    const DROP: unsafe extern "C" fn(*mut Self) = drop_rdkafka;
+}
+
+unsafe extern "C" fn drop_rdkafka(ptr: *mut RDKafka) {
+    tokio::task::block_in_place(|| rdsys::rd_kafka_destroy(ptr))
 }
 
 // The library is completely thread safe, according to the documentation.
diff --git a/src/consumer/base_consumer.rs b/src/consumer/base_consumer.rs
index 623a7127..af8e4b82 100644
--- a/src/consumer/base_consumer.rs
+++ b/src/consumer/base_consumer.rs
@@ -541,17 +541,6 @@ where
     }
 }
 
-impl<C> Drop for BaseConsumer<C>
-where
-    C: ConsumerContext,
-{
-    fn drop(&mut self) {
-        trace!("Destroying consumer: {:?}", self.client.native_ptr()); // TODO: fix me (multiple executions ?)
-        unsafe { rdsys::rd_kafka_consumer_close(self.client.native_ptr()) };
-        trace!("Consumer destroyed: {:?}", self.client.native_ptr());
-    }
-}
-
 /// A convenience iterator over the messages in a [`BaseConsumer`].
 ///
 /// Each call to [`Iter::next`] simply calls [`BaseConsumer::poll`] with an
```

The real diff would need to be ever so slightly larger, because rust-rdkafka tries to be generic over the async runtime, but that would be a relatively mechanical change.

guswynn commented:
@benesch The `rdkafka` patch would at least be bigger as we need to fixup at least `BaseProducer` and `PartitionQueue` in addition to `BaseConsumer`. Technically we should wrap all calls that have non-default `Drop` impls, as using a new one in the future that we don't yet use, but didn't wrap, is risky. One thing I like about this pr is that it localizes our reasoning about blocking to this one place, and using new apis (for the most part) will need supported added here.

Additionally, I do not think that `block_in_place` has the right semantics for us to use it. Actually, I think this is true in general for the mz codebase, but in regards to this issue:

- `block_in_place` spawns the core worker mechanism on a `spawn_blocking` thread: https://github.com/tokio-rs/tokio/blob/ec3038357f8dbbae0ab4daf55ff27937ad42abb2/tokio/src/runtime/scheduler/multi_thread/worker.rs#L398-L424. This means that other tasks and other `block_on`'s (like the coordinator) that rely on prompt polling may see latency increases, especially if the `spawn_blocking` queue is full.
- `block_in_place` just blocks https://github.com/tokio-rs/tokio/blob/ec3038357f8dbbae0ab4daf55ff27937ad42abb2/tokio/src/runtime/scheduler/multi_thread/worker.rs#L388-L392 in `timely` threads, which means `BaseConsumer`'s being dropped in `clusterd` can block progress

The way around this in the `rdkafka` crate is to do the same thing we have done here: wrap things up in `Option`'s, and `take`-and-`spawn_blocking` them in `Drop`. This is going to be likely more code than this pr, in places that are risky to change (the `rdkafka` wrapper code is quite fragile upon investigation)

benesch commented:
Sorry for not getting back sooner!

> @benesch The `rdkafka` patch would at least be bigger as we need to fixup at least `BaseProducer` and `PartitionQueue` in addition to `BaseConsumer`. 

Wait, I don't think so! The load bearing part of the above patch is amending the drop implementation for `RDKafka`. This is the base client type that underlies `BaseProducer`, `BaseConsumer`, `PartitionQueue`, etc. The reason I removed the `Drop` impl for the `BaseConsumer` is that it's not needed; `rd_kafka_destroy` will automatically call `rd_kafka_consumer_close` for consumer clients. And so removing the `Drop` impl for `BaseConsumer` means that only the `Drop` impl for `RDKafka` needs special handling.

> Technically we should wrap all calls that have non-default `Drop` impls, as using a new one in the future that we don't yet use, but didn't wrap, is risky.

I'm not sure I follow this! With the above rdkafka patch, the only `Drop` impl that needs special handling is `RDKafka`'s. Everything else just falls out naturally from that, because all the other types in rdkafka just contain references to the underlying `RDKafka`.

> Additionally, I do not think that block_in_place has the right semantics for us to use it.

I was a bit torn on that. `block_in_place` preserves the upstream semantics of "block on `Drop` until all cleanup completes", which is perhaps more upstreamable. But if we never want to block on `Drop`, we could just unconditionally `thread::spawn` the call to `rd_kafka_destroy`.

> The way around this in the `rdkafka` crate is to do the same thing we have done here: wrap things up in `Option`'s, and `take`-and-`spawn_blocking` them in `Drop`. This is going to be likely more code than this pr, in places that are risky to change 

I don't think so? Isn't it this simple?

```diff
diff --git a/src/client.rs b/src/client.rs
index d0c6d6fa..6d3952b0 100644
--- a/src/client.rs
+++ b/src/client.rs
@@ -175,7 +175,12 @@ pub struct NativeClient {
 
 unsafe impl KafkaDrop for RDKafka {
     const TYPE: &'static str = "client";
-    const DROP: unsafe extern "C" fn(*mut Self) = rdsys::rd_kafka_destroy;
+    const DROP: unsafe extern "C" fn(*mut Self) = drop_rdkafka;
+}
+
+unsafe extern "C" fn drop_rdkafka(ptr: *mut RDKafka) {
+    let ptr = ptr as usize;
+    std::thread::spawn(move || rdsys::rd_kafka_destroy(ptr as *mut RDKafka));
 }
 
 // The library is completely thread safe, according to the documentation.
diff --git a/src/consumer/base_consumer.rs b/src/consumer/base_consumer.rs
index 623a7127..af8e4b82 100644
--- a/src/consumer/base_consumer.rs
+++ b/src/consumer/base_consumer.rs
@@ -541,17 +541,6 @@ where
     }
 }
 
-impl<C> Drop for BaseConsumer<C>
-where
-    C: ConsumerContext,
-{
-    fn drop(&mut self) {
-        trace!("Destroying consumer: {:?}", self.client.native_ptr()); // TODO: fix me (multiple executions ?)
-        unsafe { rdsys::rd_kafka_consumer_close(self.client.native_ptr()) };
-        trace!("Consumer destroyed: {:?}", self.client.native_ptr());
-    }
-}
-
 /// A convenience iterator over the messages in a [`BaseConsumer`].
 ///
 /// Each call to [`Iter::next`] simply calls [`BaseConsumer::poll`] with an
```

It doesn't seem to me like there's anything to be gained by calling `spawn_blocking` instead of `thread::spawn`, because we don't want to `await` the result.

> (the `rdkafka` wrapper code is quite fragile upon investigation)

Can you say more about what seems fragile?

petrosagg commented:
@benesch my understanding of the concern, which has not been spelled out explicitly in the thread, is that your diff will run `rd_kafka_destroy` either with `block_in_place`, which preserves the current execution order but behaves badly if it happens on a timely worker, or it will run `rd_kafka_destroy` on a new thread (either with `spawn_blocking` or a normal thread) and will **immediately return** to the caller.

So the question that is unclear to me is: Are we 100% that whatever code calls the defined `DROP` function doesn't look like this?:

```
RDKafka.DROP();
// do something that is only safe if the previous `DROP` call has actually truly ran `rd_kafka_destroy` to completion. 
```

Because if code like this exists it could be doing things like freeing memory that will be accessed by that background thread we spawned

benesch commented:
> @benesch my understanding of the concern, which has not been spelled out explicitly in the thread, is that your diff will run `rd_kafka_destroy` either with `block_in_place`, which preserves the current execution order but behaves badly if it happens on a timely worker, or it will run `rd_kafka_destroy` on a new thread (either with `spawn_blocking` or a normal thread) and will **immediately return** to the caller.
> 
> So the question that is unclear to me is: Are we 100% sure that whatever code calls the defined `DROP` function doesn't look like this?:
> 
> ```
> RDKafka.DROP();
> // do something that is only safe if the previous `DROP` call has actually truly ran `rd_kafka_destroy` to completion. 
> ```
> 
> Because if code like this exists it could be doing things like freeing memory that will be accessed by that background thread we spawned

Yep! The `NativeClient` is embedded in exactly one type: `Client`.

https://github.com/MaterializeInc/rust-rdkafka/blob/8ea07c4d2b96636ff093e670bc921892aee0d56a/src/client.rs#L225-L228

There's nothing funny going on there that I'm aware of. `Client` has no custom `Drop` implementation, and neither does `NativeClient`, so all that happens when you drop a `Client` is that `NativeClient` is dropped, which calls the `NativePtr` `Drop` implementation ([link](https://github.com/MaterializeInc/rust-rdkafka/blob/8ea07c4d2b96636ff093e670bc921892aee0d56a/src/util.rs#L266-L275)), which calls the drop method.

benesch commented:
I put up https://github.com/MaterializeInc/materialize/pull/24673 with the rdkafka patch to get a CI run while we discuss.

petrosagg commented:
> There's nothing funny going on there that I'm aware of. Client has no custom Drop implementation, and neither does NativeClient, so all that happens when you drop a Client is that NativeClient is dropped, which calls the NativePtr Drop implementation (link), which calls the drop method.

Perfect, that resolves my worry! Personally I'm now convinced that the alternative approach of modifying rust-rdkafka is safe. I'll leave the final call to Gus

guswynn commented:
replaced by https://github.com/MaterializeInc/materialize/pull/24673
